src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs

```diff
@@ -34,105 +46,94 @@ public ComponentStatePersistenceManager(ILogger<ComponentStatePersistenceManager
+    /// <param name="serializationMode">The <see cref="PersistedStateSerializationMode"/> to restore the application state.</param>
     /// <returns>A <see cref="Task"/> that will complete when the state has been restored.</returns>
-    public async Task RestoreStateAsync(IPersistentComponentStateStore store)
+    public async Task RestoreStateAsync(IPersistentComponentStateStore store, PersistedStateSerializationMode serializationMode)
     {
         var data = await store.GetPersistedStateAsync();
         State.InitializeExistingState(data);
+        // We need to set the serialization in order to register callbacks later
+        State.SerializationMode = serializationMode;
     }
 
     /// <summary>
     /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>.
     /// </summary>
-    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to restore the application state from.</param>
+    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to persist the application state into.</param>
+    /// <param name="serializationMode">The <see cref="PersistedStateSerializationMode"/> to persist the application state.</param>
     /// <param name="renderer">The <see cref="Renderer"/> that components are being rendered.</param>
     /// <returns>A <see cref="Task"/> that will complete when the state has been restored.</returns>
-    public Task PersistStateAsync(IPersistentComponentStateStore store, Renderer renderer)
-        => PersistStateAsync(store, renderer.Dispatcher);
+    public Task PersistStateAsync(
+        IPersistentComponentStateStore store,
+        PersistedStateSerializationMode serializationMode,
+        Renderer renderer)
+        => PersistStateAsync(store, serializationMode, renderer.Dispatcher);
 
     /// <summary>
-    /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>.
+    /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>
+    /// so that it could be restored on Server.
     /// </summary>
-    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to restore the application state from.</param>
+    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to persist the application state into.</param>
+    /// <param name="serializationMode">The <see cref="PersistedStateSerializationMode"/> to persist the application state.</param>
     /// <param name="dispatcher">The <see cref="Dispatcher"/> corresponding to the components' renderer.</param>
     /// <returns>A <see cref="Task"/> that will complete when the state has been restored.</returns>
-    public Task PersistStateAsync(IPersistentComponentStateStore store, Dispatcher dispatcher)
+    public Task PersistStateAsync(
+        IPersistentComponentStateStore store,
+        PersistedStateSerializationMode serializationMode,
```

Comment:
Explained in this [comment](https://github.com/dotnet/aspnetcore/pull/50373#discussion_r1313171647)

Reply:
I sketched out some of my ideas in  https://github.com/dotnet/aspnetcore/commit/f7a8552d1de5076e123f48f7cf06f8a4224ac33e to show how this could be achieved 🙂